assembly/src/ast/procedure/name.rs

```diff
@@ -0,0 +1,383 @@
+    }
+}
+
+impl PartialOrd for FullyQualifiedProcedureName {
+    fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl From<FullyQualifiedProcedureName> for miette::SourceSpan {
+    fn from(fqn: FullyQualifiedProcedureName) -> Self {
+        fqn.span.into()
+    }
+}
+
+impl Spanned for FullyQualifiedProcedureName {
+    fn span(&self) -> SourceSpan {
+        self.span
+    }
+}
+
+impl fmt::Debug for FullyQualifiedProcedureName {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("FullyQualifiedProcedureName")
+            .field("module", &self.module)
+            .field("name", &self.name)
+            .finish()
+    }
+}
+
+impl fmt::Display for FullyQualifiedProcedureName {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}::{}", &self.module, &self.name)
+    }
+}
+
+/// Serialization
+impl FullyQualifiedProcedureName {
+    /// Serialize to `target` using `options`
+    pub fn write_into_with_options<W: ByteWriter>(&self, target: &mut W, options: AstSerdeOptions) {
+        if options.debug_info {
+            self.span.write_into(target);
+        }
+        self.module.write_into(target);
+        self.name.write_into_with_options(target, options);
+    }
+
+    /// Deserialize from `source` using `options`
+    pub fn read_from_with_options<R: ByteReader>(
+        source: &mut R,
+        options: AstSerdeOptions,
+    ) -> Result<Self, DeserializationError> {
+        let span = if options.debug_info {
+            SourceSpan::read_from(source)?
+        } else {
+            SourceSpan::default()
+        };
+        let module = LibraryPath::read_from(source)?;
+        let name = ProcedureName::read_from_with_options(source, options)?;
+        Ok(Self { span, module, name })
+    }
+}
+
+// PROCEDURE NAME
+// ================================================================================================
+
+/// Procedure name.
+///
+/// The symbol represented by this type must comply with the following rules:
+///
+/// - It must start with an ASCII alphabetic character, or one of: `_`, `.`, or `$`
+/// - If it starts with a non-ASCII alphabetic character, it must contain at least one ASCII
+///   alphabetic character, e.g. `_`, `$_` are not valid symbols, but `_a` or `$_a` are.
+/// - Otherwise, the name may consist of any number of printable ASCII characters, e.g.
+///   alphanumerics, punctuation, etc. Control characters and the like are explicitly not allowed.
+///
+/// NOTE: In Miden Assembly source files, a procedure name must be quoted in double-quotes if it
+/// contains any characters other than ASCII alphanumerics, `_`, or `::`. See examples
+/// below.
+///
+/// ## Examples
+///
+/// ```masm,ignore
+/// # All ASCII alphanumeric, bare identifier
+/// proc.foo
+///   ...
+/// end
+///
+/// # All ASCII alphanumeric, namespaced
+/// proc.std::foo
+///   ...
+/// end
```

Comment:
Yeah I think this could use a bit of cleanup/clarification - currently we don't actually allow `::` in bare identifiers, in order to recognize it as a namespacing operator in the parser. You can use `::` in identifiers by quoting them of course, since that is unambiguous.

However, I do want to note that I intend to revisit identifiers/symbols in Miden Assembly in a future PR, as follows:

* All symbols associated with module items occupy a single global namespace
* Imports become optional syntax sugar for aliasing externally-defined symbols in the current lexical scope with a more convenient name, i.e. `use.std` is redundant, as something like `exec.std::foo` is already fully-qualified, but `use.std::u64` would allow calling `std::u64` functions using just the `u64` alias. As long as a symbol is fully-qualified, no explicit import is required
* The `LibraryNamespace` of a symbol determines whether or not it opts-in to the syntax sugar for symbols that follow the current MASM namespacing rules:
  - A module namespaced with `LibraryNamespace::User` will have all of its exported symbols expanded to the form `<namespace>(::<module_name>)*::<export_name>`. 
  - A module namespaced with `LibraryNamespace::(Anon|Exec|Kernel)` will have no implicit namespacing, i.e. all exports would be defined in the global namespace as-is
 * Symbols defined in a module implicitly shadow the global symbol table within that module. This allows a module to contain private constants/procedures with names that might otherwise conflict with symbols in the global namespace, without any ambiguity. However, there may be times where you might need to reference a shadowed symbol, so to allow for that, the syntax is also extended to support referencing symbols in the global namespace directly, using `::<symbol>`. This would behave the same as `::<path>` in Rust, i.e. disabling any implicit scoping/shadowing so that one can reference things globally without conflict.
 
This has the following benefits/effects:

* Compilers can treat Miden Assembly like a normal assembly language target, and do not have to try and map to Miden Assembly namespacing/identifier syntax - they can simply opt-out of namespaces, and use fully-qualified identifiers everywhere.
* Hand-written Miden Assembly retains some convenient syntax sugar that makes it less verbose
* Syscalls will no longer have confusing resolution rules (i.e. currently it is the case that symbols which appear local, are actually resolved in the kernel module, _not_ the calling module, which is not how any of the other call instructions work). After these changes, syscalls would use the same resolution rules as everything else, with the exception that we would validate that the named procedure actually exists in the kernel.
* Modules in existing libraries, such as the standard library, largely require no changes as a result of this, the exception may be kernel procedures/syscalls, which may require some explicit disambiguation, but it is possible they are already unambiguous in which case no changes would be needed.
* "Private" symbols, i.e. those not exported from a module, do not require entries in the global symbol table, so they can never conflict with those of other modules. We would still check for conflicts within a module, but after local procedure references are resolved during semantic analysis, those symbols are no longer used.
* We can disambiguate symbols in the global namespace and user-defined namespaces, even if they expand to the same symbol name in theory, due to the `LibraryNamespace` component of the identifier. In the surface syntax of Miden Assembly, identifiers must be quoted to use `::` when opting-out of namespacing rules, so it can never be the case that the syntax is ambiguous either, i.e. you end up with either `"std::foo"` (no namespacing) or `std::foo` (namespaced).


This is of primary importance from the compiler end of things, especially with Rust, as it assumes that the namespace of the target assembler is flat. The changes described above will let us mix and match, i.e. emit the bulk of the code without any namespacing at all, but still retain the ability to call into hand-written MASM which does make use of namespacing. It should also mesh a bit better with the distinction between different types of modules (e.g. executable vs library).

Anyway, I'm going to write this up as a new issue along with the others, but wanted to note it here.